src/main/java/org/opensearch/knn/training/TrainingJobRunner.java

```diff
@@ -176,4 +193,106 @@ private void serializeModel(TrainingJob trainingJob, ActionListener<IndexRespons
     public int getJobCount() {
         return jobCount.get();
     }
+
+    /**
+     * This method is called whenever the cluster state changes. It is used to update models that are still training when a node leaves or the cluster crashes.
+     * @param event the event that changed the cluster change
+     */
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.localNodeClusterManager()) {
+            if (event.isNewCluster()) {
+                threadPool.schedule(() -> {
+                    try {
+                        updateModelsNewCluster();
+                    } catch (IOException | InterruptedException | ExecutionException e) {
+                        throw new RuntimeException(e);
+                    }
+                }, TimeValue.timeValueSeconds(1), ThreadPool.Names.GENERIC);
+            } else if (event.nodesRemoved()) {
+                List<DiscoveryNode> removedNodes = event.nodesDelta().removedNodes();
+                threadPool.schedule(() -> {
+                    try {
+                        updateModelsNodesRemoved(removedNodes);
+                    } catch (IOException | InterruptedException | ExecutionException e) {
+                        throw new RuntimeException(e);
+                    }
+                }, TimeValue.timeValueSeconds(0), ThreadPool.Names.GENERIC);
+            }
+        }
+    }
+
+    protected void updateModelsNewCluster() throws IOException, InterruptedException, ExecutionException {
+        if (modelDao.isCreated()) {
+            List<String> modelIds = searchModelIds();
```

Comment:
I think we should be reading modelIds from cluster metadata, which will be available on all of the nodes.

Reply:
Is the modelDao object not available on all of the nodes?  This is typically the object that is used by most of the codebase when interacting with any of the models.

Reply:
Right, modelDao will be available. In general, I want to avoid transport calls and searching indices if the data is already available on each node in the metadata.

I think instead, we should add a function in ModelDao that basically returns all models. Something like https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/indices/ModelDao.java#L449.

Reply:
Yeah that should be more efficient

Reply:
Agree with @jmazanec15. If the data is available in cluster metadata, we should try to use it as it is cheaper to retrieve the data. If you use cluster metadata, I think you can also avoid using `threaedPool.schedule()`